sycl/include/CL/sycl/exception.hpp

```diff
@@ -12,40 +12,46 @@
 
 #include <CL/sycl/detail/common.hpp>
 #include <CL/sycl/stl.hpp>
-#include <exception>
+
+#include <cstddef>
 
 namespace cl {
 namespace sycl {
 
+// Forward declaration
 class context;
 
 struct exception {
+
+  using ContextPtr = shared_ptr_class<context>;
+
   exception() = default;
 
-  const char *what() const noexcept { return msg.c_str(); }
+  exception(const char *Msg, const cl_int ClErr = CL_SUCCESS,
+            ContextPtr Context = nullptr)
+      : Msg(std::string(Msg) + " " +
+            ((ClErr == CL_SUCCESS) ? "" : OCL_CODE_TO_STR(ClErr))),
+        ClErr(ClErr), Context(Context) {}
+
+  exception(const std::string &Msg, const cl_int ClErr = CL_SUCCESS,
+            ContextPtr Context = nullptr)
+      : exception(Msg.c_str(), ClErr, Context) {}
+
+  const char *what() const noexcept { return Msg.c_str(); }
   bool has_context() const;
   context get_context() const;
   cl_int get_cl_code() const;
 
 private:
-  std::string msg = "Message not specified";
-  cl_int cl_err = CL_SUCCESS;
-  shared_ptr_class<context> Context;
-
-protected:
-  exception(const char *msg, int cl_err = CL_SUCCESS,
-            shared_ptr_class<context> Context = nullptr)
-      : msg(std::string(msg) + " " +
-            ((cl_err == CL_SUCCESS) ? "" : OCL_CODE_TO_STR(cl_err))),
-        cl_err(cl_err), Context(Context) {}
-
-  exception(const std::string &msg, int cl_err = CL_SUCCESS,
-            shared_ptr_class<context> Context = nullptr)
-      : exception(msg.c_str(), cl_err, Context) {}
+  std::string Msg = "Message not specified";
+  cl_int ClErr = CL_SUCCESS;
+  ContextPtr Context;
 };
 
 // Forward declaration
-namespace detail { class queue_impl; }
+namespace detail {
+class queue_impl;
```

Comment:
Do you need this forward declaration? `friend class detail::queue_impl` below should be a forward declaration itself.